xonsh/completer.py

```diff
@@ -396,30 +417,32 @@ def attr_complete(self, prefix, ctx):
         return attrs
 
     def _all_commands(self):
+        """Possibly update our set of cached commands and return the set."""
         path = builtins.__xonsh_env__.get('PATH', [])
-        # did PATH change?
+        # Did PATH change?
         path_hash = hash(tuple(path))
         cache_valid = path_hash == self._path_checksum
         self._path_checksum = path_hash
-        # did aliases change?
+        # Did aliases change?
         al_hash = hash(tuple(sorted(builtins.aliases.keys())))
         self._alias_checksum = al_hash
         cache_valid = cache_valid and al_hash == self._alias_checksum
         pm = self._path_mtime
-        # did the contents of any directory in PATH change?
-        for d in filter(os.path.isdir, path):
+        # Did the contents of any directory in PATH change?
+        for d in (p for p in path if os.path.isdir(p)):
```

Comment:
The answer is that in Python there should be one obvious way to do something. And the Python community has decided that is list/set/dict comprehensions. Which is why pylint complains about idioms more appropriate for scala, etc. In the Go lang world there aren't any such arguments because an deviation from the output of gofmt is deemed invalid. Pylint is the closest thing we have to that in that in the Python community.